src/shared/firewall-util-nft.c

```diff
@@ -1130,3 +1173,160 @@ int fw_nftables_add_local_dnat(
         /* table created anew; previous address already gone */
         return fw_nftables_add_local_dnat_internal(ctx->nfnl, add, af, protocol, local_port, remote, remote_port, NULL);
 }
+
+static const char *const nfproto_table[] = {
+        [NFPROTO_ARP]    = "arp",
+        [NFPROTO_BRIDGE] = "bridge",
+        [NFPROTO_INET]   = "inet",
+        [NFPROTO_IPV4]   = "ip",
+        [NFPROTO_IPV6]   = "ip6",
+        [NFPROTO_NETDEV] = "netdev",
+};
+
+DEFINE_STRING_TABLE_LOOKUP(nfproto, int);
+
+void nft_set_context_free_many(NFTSetContexts *s) {
+        assert(s);
+
+        for (size_t i = 0; i < s->n_contexts; i++) {
+                free(s->contexts[i].table);
+                free(s->contexts[i].set);
+        }
+
+        s->n_contexts = 0;
+        s->contexts = mfree(s->contexts);
+}
+
+int nft_set_context_add(NFTSetContexts *s, int nfproto, const char *table, const char *set) {
+        _cleanup_free_ char *table_dup = NULL, *set_dup = NULL;
+        assert(s);
+        assert(IN_SET(nfproto, NFPROTO_ARP, NFPROTO_BRIDGE, NFPROTO_INET, NFPROTO_IPV4, NFPROTO_IPV6, NFPROTO_NETDEV));
+        assert(table);
+        assert(set);
+
+        table_dup = strdup(table);
+        if (!table_dup)
+                return -ENOMEM;
+
+        set_dup = strdup(set);
+        if (!set_dup)
+                return -ENOMEM;
+
+        NFTSetContext *c;
+        c = greedy_realloc((void **) &s->contexts, s->n_contexts + 1, sizeof(NFTSetContext));
+        if (!c)
+                return -ENOMEM;
+
+        s->contexts = c;
+
+        c[s->n_contexts++] = (NFTSetContext) {
+                .nfproto = nfproto,
+                .table = TAKE_PTR(table_dup),
+                .set = TAKE_PTR(set_dup),
+        };
+
+        return 0;
+}
+
+int nft_set_context_dup(const NFTSetContexts *src, NFTSetContexts *dst) {
+        assert(src);
+        assert(dst);
+        int r;
+        NFTSetContexts d = (NFTSetContexts) {};
+
+        for (size_t i = 0; i < src->n_contexts; i++) {
+                r = nft_set_context_add(&d, src->contexts[i].nfproto, src->contexts[i].table, src->contexts[i].set);
+                if (r < 0) {
+                        nft_set_context_free_many(&d);
+                        return r;
+                }
+        }
+
+        dst->n_contexts = d.n_contexts;
+        dst->contexts = d.contexts;
+
+        return 0;
+}
+
+int config_parse_nft_set_context(
+                const char *unit,
+                const char *filename,
+                unsigned line,
+                const char *section,
+                unsigned section_line,
+                const char *lvalue,
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata) {
+        NFTSetContexts *nft_set_context = ASSERT_PTR(data);
```

Comment:
Please insert an empty line before variable declarations.

Reply:
Sorry, forgot to do this on this round.